Add lazy `PMap` reader

gonzaponte opened the pull request:
Cities that read `PMap`s consume a large amount of RAM because all events are kept in memory simultaneously. This PR adds new functions to read the `PMap`s on an event-by-event basis, in a lazy manner. The old readers have been renamed using the suffix `_eager` for clarity, while the new ones contain the suffix `_lazy`. The names of the old readers have been reused in the interfacing functions that allow the user to switch between the lazy and the eager readers. The readers' default to the eager mode for backward compatibility. However, the cities will read them lazily to avoid large memory allocations.

Moreover, the lazy versions contain two arguments to read a subset of all pmaps. These are `skip`, and `n`, which will skip the first `skip` events and read the following `n` events.

This PR has been tested in production mode.

gonzaponte committed:
Rename `load_pmaps` -> `load_pmaps_eager`

... and related tests

gonzaponte committed:
Add `load_pmaps{,_as_df}` as aliases to their eager versions

gonzaponte committed:
Add tests for lazy versions of `load_pmaps{,_as_df}`

gonzaponte committed:
Add lazy versions of `load_pmaps{,_as_df}`

gonzaponte committed:
Parametrize lazyness of `load_pmaps{,_as_df}`

gonzaponte committed:
Add tests for parametrized `load_pmaps{,_as_df}`

gonzaponte committed:
Use lazy pmap reader in cities

gonzaponte committed:
Check file integrity in readers

gonzaponte committed:
Fix test

Now input file must have the /Run/events node in order to check for
the file integrity. We fix the test by adding this table.

jwaiton commented on invisible_cities/io/pmaps_io_test.py:
```diff
@@ -201,6 +205,18 @@ def test_load_pmaps_as_df_lazy(KrMC_pmaps_filename, KrMC_pmaps_dfs):
+
+    assert isinstance(lazy, Generator)
+    element = next(lazy)
+    assert len(element   ) == 5
```
Is there a reason this spacing after `element` is the way it is? It doesn't appear to align with anything

gonzaponte replied:
Probably a leftover from a previous version. Will fix.

jwaiton commented on invisible_cities/io/pmaps_io_test.py:
```diff
@@ -214,14 +244,46 @@ def test_load_pmaps(output_tmpdir, data):
 
+def test_load_pmaps_lazy(KrMC_pmaps_filename):
+    """Ensure the lazy and non-lazy versions provide the same result"""
+    pmaps_eager = pmpio.load_pmaps_eager(KrMC_pmaps_filename)
+    pmaps_lazy  = pmpio.load_pmaps_lazy (KrMC_pmaps_filename)
```
Should these not follow the new format shown in the test below: `load_pmaps(..., lazy=False/True)`

gonzaponte replied:
I prefer to test the functions directly. A separate test ensures that `load_pmaps` (meant to be the interface to these functions(*)) works correctly with both values of the `lazy` parameter. But that's another layer. Testing the functions directly is more robust; in theory, something in `load_pmaps` could prevent errors in one of the inner functions from being spotted.

(*) Perhaps, we should "hide" the inner functions by prefixing an underscore to their names. This indicates the user that they should use `load_pmaps` with the corresponding value of `lazy` and not `load_pmaps_whatever`. What do you think?

jwaiton replied:
That makes sense. Testing the inner function explicitly is important :+1: 

On the (*), I don't see why not other than convention. Is the use of leading underscores implemented anywhere else in IC? For some users this may be more of a hindrance than a help if they are unfamiliar with its effects.

gonzaponte replied:
This is more of a Python convention than an IC one. Even though we don't follow PEP8 entirely, this specific rule is a common convention in the Python community. It's described [here](https://peps.python.org/pep-0008/#descriptive-naming-styles) (last paragraph).

jwaiton replied:
I think I'm being overly concerned. My worry was that (as is described in the PEP8 doc you linked) adding a leading underscore means indiscriminate importing (`from components import *`) will not import the corresponding inner function. I just assumed that could cause issues with users if it hasn't been used much in IC before and they aren't familiar with PEP8.

I think it's a good change to add the underscore, I'm just concerned in case notebooks start crashing and people aren't sure why.

gonzaponte replied:
The takeaway message here is that using this convention is just a way of expressing something. That something is that users should exclusively use the `load_pmaps` function, and not the other ones. We don't forbid their use, just discourage it. If a user insists on calling the inner functions directly, they will have to import them explicitly. Unfortunately, there is no other solution for the case you mention than learning a bit more about the language.

jwaiton commented on invisible_cities/io/pmaps_io_test.py:
```diff
@@ -214,14 +244,46 @@ def test_load_pmaps(output_tmpdir, data):
+        assert_PMap_equality(pmap_lazy, pmaps_eager[evt])
+
+
+def test_load_pmaps_eager_lazy(KrMC_pmaps_filename):
```
This test appears to be ensuring that the content from `load_pmaps` is as expected (type and length-wise), but the test name is a big vague in that sense. Especially since the previous test (`test_load_pmaps_lazy`) is almost the same name, but accomplishes a different task. Maybe this should be changes?

gonzaponte replied:
I'm struggling to find a good name. Do you have any suggestions?

jwaiton replied:
I guess `test_load_pmaps_returns_correct_format` or `...correct_type_and_shape`? The `lazy` and `eager` are both tested so I don't think they need to be specified distinctly.

jwaiton commented on invisible_cities/io/pmaps_io_test.py:
```diff
@@ -203,7 +233,7 @@ def test_load_pmaps_as_df_without_ipmt(KrMC_pmaps_without_ipmt_filename, KrMC_pm
 
 @given(stg.data())
-def test_load_pmaps(output_tmpdir, data):
+def test_load_pmaps_eager(output_tmpdir, data):
```
Likewise in naming convention for this test, its a bit vague since it matches the other related tests in naming convention but differs in content. 

It appears to be writing a pmap, then reading it in and comparing it to itself 'pre-writing'. Is that correct? In that case, is there a reason that the pmap is generated here explicitly but in the other tests `KrMC_pmaps_filename` is used?

gonzaponte replied:
Hmm, I didn't read the test, actually, I just renamed it and refactored it to make it work with the new version. But I don't see a reason why it shouldn't follow the same logic as the other ones, so I will change it.

gonzaponte replied:
I just realized that some of other tests are wrong, as they compare the output of `load_pmaps_as_df_eager` with the output of a fixture, which only reads pmaps using that very function. So it's testing nothing essentially. I will refactor this test to make it simpler (but keep the logic) and fix the other ones.

jwaiton reviewed (commented):
I've read through the PR, very interesting work! My comments at the moment are mostly cosmetic/about tests. 

I have read all the new lazy functions, but my knowledge of generators is a bit murky so I need to take more time to understand exactly how they work here in the pmap loader.

gonzaponte committed:
Add fixtures with custom pmaps

gonzaponte committed:
Rename tests and add docstring

gonzaponte committed:
Deprecate reading pmaps without per-sensor information

This might come back in the future, but it will have to be refactored
anyway.

Moreover, this test was not doing anything sensible. so it will be removed

gonzaponte committed:
Remove unused fixture from arguments

jwaiton commented on invisible_cities/io/pmaps_io.py:
```diff
@@ -78,8 +80,23 @@ def _make_tables(hdf5_file, compression):
 
 
-def load_pmaps_as_df(filename):
+def check_file_integrity(file):
+    events_run      = file.root.Run  .events.read(field="evt_number")
+    events_pmaps_s1 = file.root.PMAPS.S1    .read(field="event")
+    events_pmaps_s2 = file.root.PMAPS.S2    .read(field="event")
+    if set(events_run) != set(events_pmaps_s1.tolist() + events_pmaps_s2.tolist()):
+        raise InvalidInputFileStructure("Inconsistent data: event number mismatch")
```
This function isn't tested explicitly, but is in the `eager` and `lazy` df functions below. Should it be tested explicitly? For example, does it behave correctly when there are multiple S2 peaks per event?

gonzaponte replied:
Yes, it should be tested explicitly. I will add some tests.

I had to put this on hold for a bit, but I was trying to rewrite the tests that were "testing nothing" as we discussed in a separate comment.

gonzaponte committed:
Add tests for `check_file_integrity`

gonzaponte committed:
Rewrite `test_load_pmaps_as_df_eager`

This test tested nothing. Now at least it makes sense.

gonzaponte committed:
Rewrite `test_load_pmaps_eager`

This now includes an explicit pmap coded in a fixture instead of using
the composite implemented in a different file. Hopefully this makes
the test more clear.

gonzaponte committed:
Add `skip` and `n` parameters to lazy reader

gonzaponte committed:
Add tests for `skip`  and `n` parameters

jwaiton commented on invisible_cities/io/pmaps_io_test.py:
```diff
@@ -29,9 +32,102 @@
+    run_number    = 0 # irrelevant
+    timestamp     = 0 # irrelevant
+
+    with tb.open_file(pmap_filename, "w") as output_file:
```
A very nice set of fixtures!

If you run the tests multiple times, this method here (line 121 onwards, not sure why it cropped it like this) will overwrite the existing `"two_pmaps.h5"` file every time right? I assume its a very small gain, but might it be worth checking that the file doesnt exist before writing to it?

gonzaponte replied:
The file won't exist, because it's being written in a temp folder, which is different every time we run the tests. As you say, it's a very small file, so the time spent here over the time it takes the test suite to run is negligible.

jwaiton commented on invisible_cities/io/pmaps_io_test.py:
```diff
@@ -320,6 +320,18 @@ def test_load_pmaps_as_df_lazy(KrMC_pmaps_filename):
+@mark.parametrize("skip" , (0, 1, 2, 3))
+@mark.parametrize("nread", (0, 1, 2, 3))
+def test_load_pmaps_as_df_lazy_subset(KrMC_pmaps_filename, skip, nread):
+    """Ensure the lazy and non-lazy versions provide the same result"""
```
The docstring here implies that the `eager` version is implemented, which isn't true right? It just checks that the subset output from the lazy method has the same number of unique entries as is expected by `nread`.

gonzaponte replied:
Copy-pasting has its risks... :)

jwaiton commented on invisible_cities/io/pmaps_io_test.py:
```diff
@@ -371,6 +383,16 @@ def test_load_pmaps_lazy(KrMC_pmaps_filename):
 
+@mark.parametrize("skip" , (0, 1, 2, 3))
+@mark.parametrize("nread", (0, 1, 2, 3))
+def test_load_pmaps_lazy_subset(KrMC_pmaps_filename, skip, nread):
```
Same as comment on `test_load_pmaps_as_df_lazy_subset`, except instead it just checks that the lengths are the same.

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -567,11 +567,11 @@ def pmap_from_files(paths):
-                yield dict(pmap=pmaps[event_number], run_number=run_number,
+                if event_number != evt:
+                    raise RuntimeError("Inconsistent data: event number mismatch")
+                yield dict(pmap=pmap, run_number=run_number,
```
This check for inconsistent data is implemented, but not tested. Similar functionality is tested for `check_file_integrity`, but that only appears to be implemented for `..._as_df_...` functions. Is it possible for these to be merged in some capacity?

gonzaponte replied:
I don't know what you mean by merging here, but I'll add a test for this reader.
Regarding `check_file_integrity`, it's only implemented for the `as_df` readers because the other ones call these, so that check is performed in any case. Is that what you were referring to?

jwaiton replied:
Yes that's what I was referring to. Both errors provided the same message with different error types, so I assumed the check for `Inconsistent data` could be merged into a function. On second reading I can see that doesn't make sense, apologies.

gonzaponte replied:
I will also change the Exception to be the same in both cases.

jwaiton commented on invisible_cities/io/pmaps_io.py:
```diff
@@ -141,6 +182,19 @@ def load_pmaps(filename):
     return pmap_dict
 
 
+def load_pmaps_lazy(filename, skip=0, n=None):
```
I would add in the PR first comment a note of what these two parameters (`skip` and `n`) do for posterity, if these functions aren't going to have docstrings.

gonzaponte replied:
I'll do both (write it in the PR and add docstrings).

jwaiton reviewed (commented):
It looks good! I like the new tests and fixtures. Just a few comments.

gonzaponte committed:
Add docstrings

gonzaponte committed:
Add test for `pmap_from_files`

gonzaponte committed:
Fix copy-pasted docstrings

jwaiton reviewed (approved):
The lazy `PMap` reader allows for memory to be utilised more efficiently when reading in PMaps, and is fully tested with documentation to explain how the relevant functions work. Good job! Approved :smile:
